### ISSUE TITLE
Using templated operator== inside TEST_CASE changes deduced types of forwarding references

### ISSUE THREAD

**Issue by 60rntogo** [onqtam/doctest#399]
I was writing some generic code that among other things allows customization of operators for user provided classes. This involved a lot of templates and forwarding references, so my tests also included cases that ensured that my templates are deducing the types that I wanted. However, it looked like the behavior was subtly different between the test code and production code which was very confusing. Ultimately, I manged to reduce my working example to the following (dummy implementations for brevity):
```C++
  template <typename T>
  bool operator==(T&&, T&&)
  {
    return true;
  }

  struct X {};

  TEST_CASE("test")
  {
    X x{};
    CHECK(x == X{});
  }
```
This code compiles (doctest 2.4.0 with gcc 10.0.1), but it should not. Indeed, if I use the comparison `x == X{}` outside of `TEST_CASE`, gcc says `deduced conflicting types for parameter ‘T’ (‘X&’ and ‘X’)` which sounds exactly right to me. The same thing happens with other comparison operators like `operator<`, but nothing else as far as I can tell.

I assume that doctest does some magic behind the scenes to parse expressions inside `CHECK` which is a cause of this and I can appreciate that it might be difficult to eliminate this side effect. But it would be great if there was a way to fix this or at least have some workaround.

**Comment by onqtam** [comment]
Sorry for the late reply. This issue is not present in Catch2 and thus it should be fixed in doctest as well. However I cannot promise when... Thanks for reporting!

**Comment by onqtam** [comment]
closed thanks to #468 (merged in `dev`)
